File: server/src/services/programs.service.ts
import { query } from "../db/db";
import { Program } from "../models/Program";
import { ProgramExercise } from "../models/Exercise";

class ProgramsService {
  async getProgram(id: string) {
    const { rows } = await query(
      `--sql
      SELECT programs.id, patients.name as patient, programs.name, description FROM programs 
      JOIN patients ON patient_id = patients.id 
      WHERE programs.id = $1
      `,
      [id]
    );

    if (rows[0] == null) {
      throw "Program does not exist";
    }
    return Program.fromDb(Object.values(rows[0]));
  }

  async createProgram(
    id: string,
    name: string,
    patientId: string,
    description: string
  ) {
    await query(
      `--sql
    INSERT INTO programs (id, name, patient_id, description)
    VALUES ($1, $2, $3, $4)
    `,
      [id, name, patientId, description]
    );
  }

  async createProgramExercises(
    programId: string,
    exercises: ProgramExercise[]
  ) {
    await Promise.all(
      exercises.map((exercise) =>
        query(
          `--sql
    INSERT INTO program_exercises (id, exercise_id, program_id, notes)
    VALUES ($1, $2, $3, $4)
    `,
          [exercise.id, exercise.exerciseId, programId, exercise.notes]
        )
      )
    );
  }

  async getProgramExercises(id: string) {
    const { rows } = await query(
      `--sql
      SELECT id, exerciseId, name, mediaLink, description, notes, sets, weight, units, reps, time, hours, minutes, seconds FROM (
        SELECT program_id, exercise_modifications.exercise_id as exerciseId, program_exercises.id as id, exercises.name as name, 
        media_link as mediaLink, description, notes, time, sets, reps, weight, units, hours, minutes, seconds,
        row_number() OVER (partition by program_exercises.exercise_id ORDER BY TIME DESC) AS row_num
        FROM program_exercises 
        JOIN exercises on program_exercises.exercise_id = exercises.id 
        JOIN exercise_modifications ON exercise_modifications.exercise_id = program_exercises.id 
        ) sub
      WHERE program_id = $1 AND row_num=1
      `,
      [id]
    );
    return rows.map((r) => ProgramExercise.fromDb(Object.values(r)));
  }

  async getExerciseProgress(id: string) {
    await query(
      `--sql
      SELECT sets, reps, weight, units, hours, minutes, seconds, time 
      FROM exercise_modifications WHERE exercise_id = $1 ORDER BY TIME DESC`,
      [id]
    );
  }

  async updateExercise(exercise: ProgramExercise, newExercise = false) {
    const {
      id,
      exerciseId,
      sets,
      reps,
      weight,
      units,
      hours,
      minutes,
      seconds,
    } = exercise;
    await query(
      `--sql
    INSERT INTO exercise_modifications (exercise_id, sets, reps, weight, units, hours, minutes, seconds)
    VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
    `,
      [
        newExercise ? id : exerciseId,
        sets,
        reps,
        weight,
        units,
        hours,
        minutes,
        seconds,
      ]
    );
  }
}
export default ProgramsService;

File: mobile/lib/services/programs.service.dart
import 'package:mobile/models/exercise.dart';

import '../models/program.dart';
import './base.service.dart';

class ProgramsService extends BaseService {
  ProgramsService() : super();

  Future<Program> getProgram(String id) async {
    final res = await request("/programs/$id", Method.get);
    if (res['program'] != null) {
      final program = res['program'];
      return Program(program['id'], program['patient'], program['name'],
          program['description']);
    } else if (res['error'] != null) {
      throw res['error'];
    } else {
      throw "Failed to fetch program";
    }
  }

  Future<void> createProgram(String name, String patientId,
      List<ProgramExercise> exercises, String description) async {
    final res = await request("/programs", Method.post, data: {
      "name": name,
      "patientId": patientId,
      "exercises": exercises,
      "description": description
    });
    if (res['error'] != null) {
      throw res['error'];
    }
  }

  Future<List<ProgramExercise>> getProgramExercises(String id) async {
    final res = await request("/programs/exercises/$id", Method.get);
    if (res['exercises'] != null) {
      final exercises = (res['exercises'] as List<dynamic>)
          .map((e) => ProgramExercise(
              id: e['id'],
              exerciseId: e['exerciseId'],
              name: e['name'],
              sets: e['sets'],
              reps: e['reps'],
              weight: double.parse(e['weight']),
              units: e['units'] == 'metric' ? Units.metric : Units.imperial,
              mediaLink: e['mediaLink'],
              notes: e['notes'],
              description: e['description'],
              hours: e['hours'],
              minutes: e['minutes'],
              seconds: e['seconds'],
              isDone: false))
          .toList();
      return exercises;
    } else if (res['error'] != null) {
      throw res['error'];
    } else {
      throw "Failed to fetch program exercises";
    }
  }

  Future<void> editProgramExercise(ProgramExercise exercise) async {
    await request("/programs/exercises/", Method.put, data: exercise.toJson());
  }
}
